tools/shared/src/main/scala/org/scalajs/core/tools/sem/Semantics.scala

```diff
@@ -107,21 +119,98 @@ final class Semantics private (
     new Semantics(
         asInstanceOfs = asInstanceOfs,
         arrayIndexOutOfBounds = arrayIndexOutOfBounds,
         moduleInit = moduleInit,
         strictFloats = strictFloats,
         productionMode = productionMode,
-        runtimeClassName = runtimeClassName)
+        runtimeClassNameMapper = runtimeClassNameMapper)
   }
 }
 
 object Semantics {
   private val HashSeed =
     scala.util.hashing.MurmurHash3.stringHash(classOf[Semantics].getName)
 
+  sealed abstract class RuntimeClassNameMapper {
+    import RuntimeClassNameMapper._
+
+    def andThen(that: RuntimeClassNameMapper): RuntimeClassNameMapper = {
+      require(!that.isInstanceOf[Custom],
+          "RuntimeClassNameMapper.custom(...) is not a valid argument to " +
+          "RuntimeClassNameMapper#andThen(), because it takes a LinkedClass " +
+          "as input instead of a String.")
+      AndThen(this, that)
+    }
+
+    private[tools] def apply(linkedClass: LinkedClass): String = {
+      def rec(mapper: RuntimeClassNameMapper, className: String): String = {
+        mapper match {
+          case KeepAll =>
+            className
+          case DiscardAll =>
+            ""
+          case ReplacePrefix(from, to) =>
+            if (className.startsWith(from))
+              to + className.substring(from.length)
+            else
+              className
+          case AndThen(first, second) =>
+            rec(second, rec(first, className))
+          case Custom(mapper) =>
+            /* Discards `className`, but that's fine because we cannot
+             * construct an AndThen(_, Custom()).
+             */
+            mapper(linkedClass)
+        }
+      }
+
+      rec(this, linkedClass.fullName)
+    }
+  }
+
+  object RuntimeClassNameMapper {
+    private case object KeepAll extends RuntimeClassNameMapper
+
+    private case object DiscardAll extends RuntimeClassNameMapper
+
+    private case class ReplacePrefix(from: String, to: String)
+        extends RuntimeClassNameMapper
+
+    private case class AndThen(first: RuntimeClassNameMapper,
+        second: RuntimeClassNameMapper)
+        extends RuntimeClassNameMapper
+
+    /** For compatibility with `RuntimeClassNameFunction`s only. */
+    private case class Custom(mapper: LinkedClass => String)
+        extends RuntimeClassNameMapper {
+      /* For compatibility of `Semantics.==` of previous versions, we need to
+       * consider all `Custom` instances as being equal, even though this
+       * definitely breaks the case class contract.
+       * Since this is deprecated, this issue will eventually go away.
+       */
+
+      override def equals(that: Any): Boolean = that.isInstanceOf[Custom]
+
+      override def hashCode(): Int = 369581025 // generated at random
+    }
+
+    def keepAll(): RuntimeClassNameMapper = KeepAll
+
+    def discardAll(): RuntimeClassNameMapper = DiscardAll
+
+    def replacePrefix(from: String, to: String): RuntimeClassNameMapper =
+      ReplacePrefix(from, to)
```

Comment:
How did you come up with these things? I'm just wondering if, for example, discarding everything with a prefix is a use-case as well.